flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1129,6 +1157,9 @@ void failGlobalIfExecutionIsStillRunning(Throwable cause, ExecutionAttemptID fai
 	 * @param t The exception that caused the failure.
 	 */
 	public void failGlobal(Throwable t) {
+		if (!isLegacyScheduling()) {
+			ExceptionUtils.rethrow(t);
```

Comment:
Re-throwing this error may not work in some cases.
`failGlobal` is used as a safety net to recover the job from unexpected inconsistent state. And in my mind this method is not intended to throw any exceptions and many invocations of it do not handle exceptions from it.

One problem I can identify is that the `CheckpointFailureManager` would break with this change as it depends on failGlobal.
Another possible issue is that all the errors happening in execution updates from TM previously lead to `failGlobal` to recover. Now the errors will cause the RPC to return an error response.

Reworking all the possible (with DefaultScheduler) entrances to `failGlobal` still requires a way to handle the errors.
Maybe we need to implement a `failGlobal` mechanism that works for DefaultScheduler.

Reply:
You are right, this deserves more attention. 😵

Reply:
> Maybe we need to implement a failGlobal mechanism that works for DefaultScheduler.

I think that's reasonable and it should be straightforward to do this. 

Moreover, I think we should differentiate calls to `failGlobal()`. The original contract for `failGlobal()` was that it is called when [_"the consistency of the execution graph cannot be guaranteed anymore"_](https://github.com/apache/flink/blob/8155d465520c4b616866d35395c3b10f7e809b78/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java#L1124). Clearly this does not hold anymore. One example that you have alreayd mentioned is the [`CheckpointFailureManager`](https://github.com/apache/flink/blob/8155d465520c4b616866d35395c3b10f7e809b78/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java#L582) – here calls to `failGlobal()` do not indicate consistency issues. For real consistency issues caused by bugs, illegal state transitions, etc. I think it is reasonable to terminate the JVM. One could propagate the exception to the JobMaster and in the RPC framework check for unwanted uncaught exceptions in specially annotated RPC methods.

Reply:
>  For real consistency issues caused by bugs, illegal state transitions, etc. I think it is reasonable to terminate the JVM.

I think it's reasonable but I'm not sure whether this change can cause much more JM failovers than it used to be. This could be an even severe problem for session cluster which holds multiple jobs in a Dispatcher.

> One could propagate the exception to the JobMaster and in the RPC framework check for unwanted uncaught exceptions in specially annotated RPC methods.

If we'd like to terminate the JVM, why not do it right in the place the fatal error happens?

Reply:
> If we'd like to terminate the JVM, why not do it right in the place the fatal error happens?

- It's easier to write unit test that assert the presence of an exception rather than termination of the JVM
- There can be unforeseeable exceptions caused by bugs (e.g., NPE, `checkState` violations), which should be handled the same way

Reply:
Let's fix this issue in a follow-up.

Reply:
Ok.

Reply:
A [ticket](https://issues.apache.org/jira/browse/FLINK-14232) is opened to track this issue.

Reply:
Done